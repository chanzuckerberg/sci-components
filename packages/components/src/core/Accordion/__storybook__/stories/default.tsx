import { Args } from "@storybook/react";
import RawAccordion, {
  AccordionDetails,
  AccordionHeader,
} from "src/core/Accordion";
import { EXTRA_SHORT_LOREM_IPSUM } from "src/common/storybook/loremIpsum";

export const Accordion = (props: Args): JSX.Element => {
  const { id, subtitle, useDivider, togglePosition } = props;
  return (
    <RawAccordion
      id={id}
      useDivider={useDivider}
      togglePosition={togglePosition}
      {...props}
    >
      <AccordionHeader id={`${id}-header`} subtitle={subtitle}>
        Accordion Header
      </AccordionHeader>
      <AccordionDetails>{EXTRA_SHORT_LOREM_IPSUM}</AccordionDetails>
    </RawAccordion>
  );
};
