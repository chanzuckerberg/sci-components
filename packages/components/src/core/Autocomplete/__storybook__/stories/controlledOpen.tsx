import { Args } from "@storybook/types";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useState } from "react";
import RawAutocomplete from "src/core/Autocomplete";
import { AUTOCOMPLETE_WRAPPER_STYLES } from "../constants";

export const ControlledOpenDemo = <T extends DefaultAutocompleteOption>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
  } = props;

  const [open, setOpen] = useState(true);

  return (
    <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
      <RawAutocomplete
        onClickAway={() => {
          setOpen(false);
        }}
        onClick={() => {
          if (open) {
            setOpen(false);
          } else {
            setOpen(true);
          }
        }}
        open={open}
        id="autocomplete-demo"
        disableCloseOnSelect={multiple}
        label={label}
        multiple={multiple}
        options={options}
        getOptionDisabled={(option: T) => {
          return option.name === "Type: feature request";
        }}
        {...props}
      />
    </div>
  );
};
