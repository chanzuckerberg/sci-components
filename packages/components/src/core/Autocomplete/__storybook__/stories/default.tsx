import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { SyntheticEvent, useEffect, useState } from "react";
import RawAutocomplete, {
  AutocompleteMultiColumnValue,
  SDSAutocompleteValue,
} from "src/core/Autocomplete";
import TagFilter from "src/core/TagFilter";
import { AutocompleteValue } from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";

export const Autocomplete = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    value: propValue,
    keepSearchOnSelect,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const isControlled = propValue !== undefined;
  const [value, setValue] = useState<
    SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(
    (multiple ? [] : null) as SDSAutocompleteValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >
  );

  const [selection, setSelection] = useState<string[]>([]);

  useEffect(() => {
    if (isControlled) {
      setValue(propValue);
    }
  }, [isControlled, propValue]);

  useEffect(() => {
    setSelection([]);
    setValue(
      (multiple ? [] : null) as SDSAutocompleteValue<
        T,
        Multiple,
        DisableClearable,
        FreeSolo
      >
    );
  }, [multiple, options]);

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocomplete
          id="autocomplete-demo"
          disableCloseOnSelect={multiple}
          label={label}
          multiple={multiple}
          onChange={handleChange}
          keepSearchOnSelect={keepSearchOnSelect}
          options={options}
          search={search}
          value={value}
          getOptionDisabled={(option: T) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleChange(
    _: SyntheticEvent<Element, Event>,
    newValue: SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setValue(newValue);
    const newSelection: string[] = [];

    if (
      (newValue as T)?.name ||
      (Array.isArray(newValue) && (newValue as T[])[0].name)
    ) {
      if (multiple) {
        setSelection(
          Array.isArray(newValue) ? newValue?.map((item) => item.name) : []
        );
      } else {
        if (newValue && !Array.isArray(newValue) && newValue.name) {
          setSelection([(newValue as T).name]);
        }
      }
    } else {
      if (multiple) {
        if (newValue) {
          Object.values(newValue).forEach((items) => {
            (items as T[])?.map(({ name }: { name: string }) => {
              newSelection.push(name);
            });
          });
          setSelection(newSelection);
        }
      } else {
        if (newValue) {
          Object.values(newValue).forEach((item) => {
            newSelection.push((item as T).name);
          });
        }
        setSelection(newSelection);
      }
    }
  }

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleTagDeleteMultiColumn(tag: string) {
    if (multiple) {
      if (value) {
        Object.entries(value).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (value) {
        Object.entries(value).forEach(([key, val]) => {
          if ((val as T).name === tag) {
            const finalValue = { ...value } as AutocompleteMultiColumnValue<
              T,
              Multiple,
              DisableClearable,
              FreeSolo
            >;
            if (finalValue) delete finalValue[key];
            setValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }

  function handleTagDelete(tag: string) {
    if (
      (value && (value as T).name) ||
      (Array.isArray(value) && (value as T[])[0].name)
    ) {
      if (multiple && Array.isArray(value)) {
        const index = value?.findIndex((item) => item.name === tag);
        const newValue = [...value];
        newValue.splice(index, 1);
        setValue(
          newValue as SDSAutocompleteValue<
            T,
            Multiple,
            DisableClearable,
            FreeSolo
          >
        );

        const newSelection = [...selection];
        const deleteIndex = newSelection.indexOf(tag);
        newSelection.splice(deleteIndex, 1);
        setSelection(newSelection);
      } else {
        setValue(
          null as SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
        );
        setSelection([]);
      }
    } else {
      handleTagDeleteMultiColumn(tag);
    }
  }
};
