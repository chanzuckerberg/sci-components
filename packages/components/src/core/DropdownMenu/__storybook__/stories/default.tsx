import { Args } from "@storybook/react";
import React, { useEffect, useRef } from "react";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import RawDropdownMenu from "src/core/DropdownMenu";
import { DROPDOWN_MENU_POPPER_POSITION } from "../constants";

export const DropdownMenu = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
  DisableClearable extends boolean | undefined,
  FreeSolo extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    title,
  } = props;
  const anchorRef = useRef(null);
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  useEffect(() => {
    setTimeout(() => {
      setAnchorEl(anchorRef.current);
    }, 1);
  }, [anchorRef]);

  return (
    <div style={{ margin: "16px 0 0 24px" }} ref={anchorRef}>
      {anchorEl ? (
        <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
          label={label}
          anchorEl={anchorEl}
          disableCloseOnSelect={false}
          multiple={multiple}
          options={options}
          PopperBaseProps={{
            disablePortal: false,
            placement: DROPDOWN_MENU_POPPER_POSITION,
          }}
          search={search}
          title={title}
          getOptionDisabled={(option: T) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      ) : null}
    </div>
  );
};
