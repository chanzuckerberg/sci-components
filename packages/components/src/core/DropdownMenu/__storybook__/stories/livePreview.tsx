import { AutocompleteValue } from "@mui/base";
import { useState } from "react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { StyledInputDropdownLive1, StyledInputDropdownLive3 } from "../style";
import {
  DROPDOWN_MENU_LIVE_PREVIEW_LABELS,
  DROPDOWN_MENU_LIVE_PREVIEW_ROW_STYLES,
  DROPDOWN_MENU_POPPER_POSITION,
  DROPDOWN_MENU_POPPER_WIDTH,
} from "../constants";
import RawDropdownMenu from "src/core/DropdownMenu";
import ButtonIcon from "src/core/ButtonIcon";

export const LivePreviewDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
  DisableClearable extends boolean | undefined,
  FreeSolo extends boolean | undefined,
>(): JSX.Element => {
  const options = DROPDOWN_MENU_LIVE_PREVIEW_LABELS;

  const [anchorEl1, setAnchorEl1] = useState<HTMLElement | null>(null);
  const [anchorEl2, setAnchorEl2] = useState<HTMLElement | null>(null);
  const [anchorEl3, setAnchorEl3] = useState<HTMLElement | null>(null);
  const [anchorEl4, setAnchorEl4] = useState<HTMLElement | null>(null);

  const [open1, setOpen1] = useState(false);
  const [open2, setOpen2] = useState(false);
  const [open3, setOpen3] = useState(false);
  const [open4, setOpen4] = useState(false);

  const [value1, setValue1] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(false));
  const [value2, setValue2] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(false));

  const [pendingValue3, setPendingValue3] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(true));
  const [pendingValue4, setPendingValue4] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(true));

  return (
    <div style={DROPDOWN_MENU_LIVE_PREVIEW_ROW_STYLES as React.CSSProperties}>
      <div style={{ alignSelf: "end" }}>
        <StyledInputDropdownLive1
          aria-describedby="live1"
          onClick={handleClick1}
          label="Click Target"
          sdsStage={open1 ? "userInput" : "default"}
          sdsType="label"
          multiple={false}
          sdsStyle="minimal"
        />

        {anchorEl1 ? (
          <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
            label="Search"
            anchorEl={anchorEl1}
            open={!!open1}
            onChange={handleChange1}
            disableCloseOnSelect={false}
            options={options.slice(0, 3) as T[]}
            PopperBaseProps={{
              placement: DROPDOWN_MENU_POPPER_POSITION,
            }}
            search={false}
            multiple={false as Multiple}
            value={value1}
            onClickAway={handleClickAway1}
            width={DROPDOWN_MENU_POPPER_WIDTH}
          />
        ) : null}
      </div>

      <div>
        <ButtonIcon
          aria-describedby="live2"
          aria-label="Open menu"
          onClick={handleClick2}
          on={false}
          sdsSize="large"
          sdsType="secondary"
          icon="InfoSpeechBubble"
        />

        {anchorEl2 ? (
          <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
            label="Search"
            anchorEl={anchorEl2}
            open={!!open2}
            search={false}
            multiple={false as Multiple}
            onChange={handleChange2}
            disableCloseOnSelect={false}
            options={options.slice(0, 3) as T[]}
            PopperBaseProps={{
              placement: DROPDOWN_MENU_POPPER_POSITION,
            }}
            value={value2}
            title="Title Lorem Ipsum"
            onClickAway={handleClickAway2}
            width={DROPDOWN_MENU_POPPER_WIDTH}
          />
        ) : null}
      </div>

      <div>
        <StyledInputDropdownLive3
          aria-describedby="live3"
          onClick={handleClick3}
          label="Click Target"
          sdsStage={open3 ? "userInput" : "default"}
          sdsType="label"
          multiple
          sdsStyle="rounded"
        />

        {anchorEl3 ? (
          <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
            label="Search"
            anchorEl={anchorEl3}
            open={!!open3}
            search
            multiple={true as Multiple}
            onChange={handleChange3}
            disableCloseOnSelect
            options={options as T[]}
            PopperBaseProps={{
              placement: DROPDOWN_MENU_POPPER_POSITION,
            }}
            value={pendingValue3}
            onClickAway={handleClickAway3}
            width={DROPDOWN_MENU_POPPER_WIDTH}
          />
        ) : null}
      </div>

      <div>
        <StyledInputDropdownLive3
          aria-describedby="live4"
          onClick={handleClick4}
          label="Click Target"
          sdsStage={open4 ? "userInput" : "default"}
          sdsType="label"
          multiple
          sdsStyle="square"
        />

        {anchorEl4 ? (
          <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
            label="Search"
            anchorEl={anchorEl4}
            open={!!open4}
            search={false}
            multiple={true as Multiple}
            groupBy={(option) => option.section as string}
            onChange={handleChange4}
            disableCloseOnSelect
            options={options as T[]}
            PopperBaseProps={{
              placement: DROPDOWN_MENU_POPPER_POSITION,
            }}
            value={pendingValue4}
            onClickAway={handleClickAway4}
            width={DROPDOWN_MENU_POPPER_WIDTH}
          />
        ) : null}
      </div>
    </div>
  );

  function getInitialValue(
    multiple: boolean
  ): AutocompleteValue<T, Multiple, DisableClearable, FreeSolo> {
    return multiple
      ? ([] as unknown as AutocompleteValue<
          T,
          Multiple,
          DisableClearable,
          FreeSolo
        >)
      : (null as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>);
  }

  function handleClickAway1() {
    setOpen1(false);
  }

  function handleClick1(event: React.MouseEvent<HTMLElement>) {
    if (open1) {
      setOpen1(false);

      if (anchorEl1) {
        anchorEl1.focus();
      }

      setAnchorEl1(null);
    } else {
      setAnchorEl1(event.currentTarget);
      setOpen1(true);
    }
  }

  function handleChange1(
    _: React.SyntheticEvent<Element, Event>,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setOpen1(false);
    setValue1(
      newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
    );
  }

  function handleClickAway2() {
    setOpen2(false);
  }

  function handleClick2(event: React.MouseEvent<HTMLElement>) {
    if (open2) {
      setOpen2(false);

      if (anchorEl2) {
        anchorEl2.focus();
      }

      setAnchorEl2(null);
    } else {
      setAnchorEl2(event.currentTarget);
      setOpen2(true);
    }
  }

  function handleChange2(
    _: React.SyntheticEvent<Element, Event>,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setOpen2(false);
    setValue2(
      newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
    );
  }

  function handleClickAway3() {
    setOpen3(false);
  }

  function handleClick3(event: React.MouseEvent<HTMLElement>) {
    if (open3) {
      setOpen3(false);

      if (anchorEl3) {
        anchorEl3.focus();
      }

      setAnchorEl3(null);
    } else {
      setAnchorEl3(event.currentTarget);
      setOpen3(true);
    }
  }

  function handleChange3(
    _: React.SyntheticEvent<Element, Event>,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    return setPendingValue3(
      newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
    );
  }

  function handleClickAway4() {
    setOpen4(false);
  }

  function handleClick4(event: React.MouseEvent<HTMLElement>) {
    if (open4) {
      setOpen4(false);

      if (anchorEl4) {
        anchorEl4.focus();
      }

      setAnchorEl4(null);
    } else {
      setAnchorEl4(event.currentTarget);
      setOpen4(true);
    }
  }

  function handleChange4(
    _: React.SyntheticEvent<Element, Event>,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    return setPendingValue4(
      newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
    );
  }
};
