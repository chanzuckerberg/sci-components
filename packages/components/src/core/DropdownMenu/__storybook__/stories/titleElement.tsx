import { AutocompleteValue } from "@mui/base";
import { Args } from "@storybook/react";
import { useEffect, useRef, useState } from "react";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { DefaultAutocompleteOption } from "src/core/Autocomplete";
import Button from "src/core/Button";
import RawDropdownMenu from "src/core/DropdownMenu";

export const TitleElementDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
  DisableClearable extends boolean | undefined,
  FreeSolo extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search = false,
  } = props;

  const anchorRef = useRef(null);
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);

  const [value, setValue] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(false));

  const [pendingValue, setPendingValue] = useState<
    AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(getInitialValue(true));

  useEffect(() => {
    setAnchorEl(anchorRef.current);
  }, []);

  const CustomHeaderComponent = () => (
    <Button sdsStyle="minimal" sdsType="primary">
      Click Me!
    </Button>
  );

  return (
    <div style={{ margin: "16px 0 0 24px" }} ref={anchorRef}>
      {anchorEl ? (
        <RawDropdownMenu<T, Multiple, DisableClearable, FreeSolo>
          {...props}
          label="Search"
          anchorEl={anchorEl}
          open
          search={search}
          multiple={multiple}
          headerComponentSlot={CustomHeaderComponent()}
          value={multiple ? pendingValue : value}
          onChange={handleChange}
          disableCloseOnSelect={multiple}
          options={options}
          onClickAway={handleClickAway}
          groupBy={(option: T) => option.section as string}
          width={300}
        />
      ) : null}
    </div>
  );

  function handleClickAway() {
    return multiple && setValue(pendingValue);
  }

  function handleChange(
    _: React.ChangeEvent<unknown>,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    if (!multiple) {
      setValue(
        newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
      );
    }

    return setPendingValue(
      newValue as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
    );
  }

  // eslint-disable-next-line sonarjs/no-identical-functions
  function getInitialValue(
    isMultiple: boolean
  ): AutocompleteValue<T, Multiple, DisableClearable, FreeSolo> {
    return isMultiple
      ? ([] as unknown as AutocompleteValue<
          T,
          Multiple,
          DisableClearable,
          FreeSolo
        >)
      : (null as AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>);
  }
};
