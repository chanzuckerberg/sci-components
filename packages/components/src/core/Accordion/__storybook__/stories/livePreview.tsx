import { Args } from "@storybook/react";
import { Accordion } from "./default";

const livePreviewStyles = {
  display: "grid",
  gridColumnGap: "24px",
  gridRowGap: "50px",
  gridTemplateColumns: "repeat(4, 200px)",
};

export function LivePreviewDemo(props: Args): JSX.Element {
  return (
    <div style={livePreviewStyles as React.CSSProperties}>
      <Accordion
        id="accordion-1"
        togglePosition="right"
        useDivider
        {...props}
      />
      <Accordion
        id="accordion-2"
        togglePosition="right"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
      <Accordion id="accordion-3" togglePosition="left" useDivider {...props} />
      <Accordion
        id="accordion-4"
        togglePosition="left"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
    </div>
  );
}
