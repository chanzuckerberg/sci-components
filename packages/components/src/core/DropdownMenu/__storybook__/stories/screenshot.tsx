import { Args } from "@storybook/react";
import Tag from "src/core/Tag";
import { DROPDOWN_MENU_LIVE_PREVIEW_LABELS } from "../constants";
import { DropdownMenu } from "./default";
import { DefaultAutocompleteOption } from "src/core/Autocomplete";

export const ScreenshotTestDemo = (props: Args): JSX.Element => {
  const TITLE_OPTIONS = [undefined, "Sample title text"];
  const SEARCH_OPTIONS = [false, true];
  const GROUP_BY_OPTIONS = [false, true];
  const SCREENSHOT_TEST_TAG_COMPONENT = {
    component: (
      <div>
        Available Labels:
        <div style={{ marginTop: 10 }}>
          <Tag
            label="bug"
            sdsStyle="rounded"
            sdsType="secondary"
            color="negative"
          />
          <Tag
            label="feature"
            sdsStyle="rounded"
            sdsType="secondary"
            color="notice"
          />
          <Tag
            label="refactor"
            sdsStyle="rounded"
            sdsType="secondary"
            color="neutral"
          />
        </div>
      </div>
    ),
    name: "custom 2",
    section: "custom component",
  };
  const SCREENSHOT_TEST_OPTIONS = DROPDOWN_MENU_LIVE_PREVIEW_LABELS.concat(
    SCREENSHOT_TEST_TAG_COMPONENT
  );

  const DISPLAY_CONTENTS: React.CSSProperties = {
    display: "contents",
  };
  const MID_LABEL: React.CSSProperties = {
    borderStyle: "solid none none none",
    gridColumn: "1 / 3",
    justifySelf: "stretch",
    paddingTop: 10,
  };

  // loop through all TITLE_OPTIONS
  return (
    <>
      {TITLE_OPTIONS.map((title, index) => {
        return <DropdownMenuTitle title={title} key={String(title) + index} />;
      })}
    </>
  );

  // loop through all SEARCH_OPTIONS + create headers for TITLE_OPTIONS
  function DropdownMenuTitle({
    title,
  }: {
    title: (typeof TITLE_OPTIONS)[number];
  }) {
    const LEVEL_STYLE: React.CSSProperties = {
      columnGap: "25px",
      display: "inline-grid",
      fontFamily: "sans-serif",
      marginLeft: "50px",
    };
    const LABEL_STYLE: React.CSSProperties = {
      fontSize: "2em",
      gridColumn: "1 / 3",
      marginBottom: 10,
    };
    return (
      <div style={LEVEL_STYLE}>
        <p style={LABEL_STYLE}>
          Title: <b>{title ? "yes" : "no"}</b>
        </p>
        {SEARCH_OPTIONS.map((search) => {
          return (
            <DropdownMenuSearch
              title={title}
              search={search}
              key={String(search)}
            />
          );
        })}
      </div>
    );
  }

  function DropdownMenuSearch({
    title,
    search,
  }: {
    title: (typeof TITLE_OPTIONS)[number];
    search: (typeof SEARCH_OPTIONS)[number];
  }) {
    const SEARCH_LABEL: React.CSSProperties = {
      ...MID_LABEL,
      borderWidth: "2px",
      fontSize: "1.17em",
      margin: 0,
    };
    const LEVEL_STYLE: React.CSSProperties = {
      margin: "0 300px 475px 0",
    };
    const GROUP_BY_LABEL: React.CSSProperties = {
      fontSize: "0.67em",
      margin: "10px 0 475px",
    };
    return (
      <div style={DISPLAY_CONTENTS}>
        <p style={SEARCH_LABEL}>
          Search: <b>{search ? "yes" : "no"}</b>
        </p>
        {GROUP_BY_OPTIONS.map((groupBy, index) => {
          const groupByFinal = groupBy
            ? (option: (typeof SCREENSHOT_TEST_OPTIONS)[number]) =>
                option.section as string
            : undefined;
          return (
            <div style={LEVEL_STYLE} key={`level-${index})}`}>
              <>
                <p style={GROUP_BY_LABEL}>
                  Grouped: <b>{groupBy ? "yes" : "no"}</b>
                </p>
                <DropdownMenu
                  {...props}
                  groupBy={groupByFinal}
                  options={
                    SCREENSHOT_TEST_OPTIONS as DefaultAutocompleteOption[]
                  }
                  title={title}
                  search={search}
                  key={String(groupBy)}
                  open
                  width={250}
                />
              </>
            </div>
          );
        })}
      </div>
    );
  }
};
