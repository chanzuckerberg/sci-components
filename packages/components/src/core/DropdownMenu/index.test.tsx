import { generateSnapshots } from "@chanzuckerberg/story-utils";
import { composeStory } from "@storybook/react";
import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import * as snapshotTestStoryFile from "./index.stories";
import Meta, { Test as TestStory } from "./index.stories";

const Test = composeStory(TestStory, Meta);

describe("<DropdownMenu />", () => {
  generateSnapshots(snapshotTestStoryFile);

  it("renders DropdownMenu component", () => {
    render(<Test {...Test.args} />);
    const DropdownMenuElement = screen.getByTestId("dropdown-menu");
    expect(DropdownMenuElement).not.toBeNull();
  });

  it("opens the menu on click", () => {
    render(<Test {...Test.args} />);
    const InputDropdownElement = screen.getByTestId("dropdown-menu");
    fireEvent.click(InputDropdownElement);
    expect(screen.getAllByText("Menu Item 1")).not.toBeNull();
  });
});
