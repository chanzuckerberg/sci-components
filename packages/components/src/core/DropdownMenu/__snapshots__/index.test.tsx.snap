// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownMenu /> Default story renders snapshot 1`] = `
<div
  style="margin: 16px 0px 0px 24px;"
/>
`;

exports[`<DropdownMenu /> MultiColumn story renders snapshot 1`] = `
<div
  style="margin: 16px 0px 0px 24px;"
/>
`;
